src/LanguageServer/Protocol/Handler/PullHandlers/VersionedPullCache.CacheItem.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
+
+internal abstract partial class VersionedPullCache<TCheapVersion, TExpensiveVersion, TState, TComputedData>
+{
+    /// <summary>
+    /// Internal cache item that updates state for a particular <see cref="Workspace"/> and <see cref="ProjectOrDocumentId"/> in <see cref="VersionedPullCache{TCheapVersion, TExpensiveVersion, TState, TComputedData}"/>
+    /// This type ensures that the state for a particular key is never updated concurrently for the same key (but different key states can be concurrent).
```

Comment:
prior to this change, there was a single semaphore around updating resultIds for the entire cache item.  However now we need to also run the computation to determine if and how the cache should be updated.  I did not want to block requests for other documents on computation of diagnostics for a different diagnostic.  So now we have the `CacheItem` which gives a finer grained semaphore to an entry for a specific workspace+proj/doc